File: local_backend/timeline.py
"""Map a saved extended cut back to its unchanged source timeline."""
from copy import deepcopy


def output_to_source(value, insertions):
    offset = 0.0
    for pause in sorted(insertions, key=lambda item: item['output_start']):
        if value < pause['output_start']:
            break
        if value < pause['output_end']:
            return pause['source_time']
        offset += pause['duration']
    return max(0.0, value - offset)


def cues_to_source(cues, insertions):
    result = []
    for cue in cues:
        item = deepcopy(cue)
        item['start'] = output_to_source(cue['start'], insertions)
        item['end'] = output_to_source(cue['end'], insertions)
        if item['end'] > item['start']:
            result.append(item)
    return result

File: dashboard/frontend/src/components/workspace/timeline.ts
import type { TimelineInsertion, TranscriptCue } from '../../localWorkspaceApi';

export function toSourceTime(value: number, insertions: TimelineInsertion[] = []) {
  let offset = 0;
  for (const pause of [...insertions].sort((a, b) => a.output_start - b.output_start)) {
    if (value < pause.output_start) break;
    if (value < pause.output_end) return pause.source_time;
    offset += pause.duration;
  }
  return Math.max(0, value - offset);
}

export function toOutputTime(value: number, insertions: TimelineInsertion[] = []) {
  return value + insertions.reduce((sum, pause) => sum + (value >= pause.source_time ? pause.duration : 0), 0);
}

export function sourceCues(cues: TranscriptCue[], insertions: TimelineInsertion[] = []) {
  return cues.map(cue => ({ ...cue, start: toSourceTime(cue.start, insertions), end: toSourceTime(cue.end, insertions) })).filter(cue => cue.end > cue.start);
}
